### PR TITLE
feat(service): Add DingTalk

### PR DESCRIPTION
**Pull request by doersa**
Missing support for the popular messaging platform DingTalk.

### CODE REVIEW

File: service/dingding/README.md [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+### 使用说明
+```go
+cfg := Config{
+  token:  "dddd",
+  secret: "xxx",
+}
+
+s := New(&cfg)
+s.Send(context.Background(), "subject", "content")
+```
```
Comment (nikoksr) [inline service/dingding/README.md:18]: Thanks for adding a readme, but I would like to stick with English as the main language for documentation and communication in this repo. However, you are of course very welcome to add something like a `README-CH.md` (assuming this is Chinese) or similar in addition to an English `README.md`.

File: service/dingding/dingding.go [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/blinkbean/dingtalk"
+	"github.com/pkg/errors"
+)
+
+type Service struct {
```
Comment (nikoksr) [inline service/dingding/dingding.go:11]: Exported struct, please add a descriptive comment.

File: service/dingding/dingding.go [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+	config Config
+	client *dingtalk.DingTalk
+}
+
+type Config struct {
```
Comment (nikoksr) [inline service/dingding/dingding.go:16]: Exported struct, please add a descriptive comment.

File: service/dingding/dingding.go [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+	Token  string
+	Secret string
+}
+
+func New(cfg *Config) *Service {
```
Comment (nikoksr) [inline service/dingding/dingding.go:21]: Exported function, please add a descriptive comment.

File: service/dingding/dingding.go [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+	return &s
+}
+
+// Send takes a message subject and a message content and sends them to all previously set users.
+func (s *Service) Send(ctx context.Context, subject, content string) error {
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	default:
+		text := fmt.Sprintf("%s\n%s", subject, content)
```
Comment (nikoksr) [inline service/dingding/dingding.go:36]:
```suggestion
		text := subject + "\n" + content
```

Please use regular string concatenation instead of `fmt.Sprintf`.

File: service/dingding/dingding_test.go [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+		content: "test content abcd",
+	}
+	s := New(&cfg)
+
+	t.Run("send", func(t *testing.T) {
+		if err := s.Send(context.Background(), arg.subject, arg.content); err != nil {
+			t.Errorf("Send() error = %v, wantErr %v", err, true)
+		}
+	})
+}
```
Comment (nikoksr) [inline service/dingding/dingding_test.go:30]: I appreciate you adding tests but as discussed [here](https://github.com/nikoksr/notify/issues/1#issuecomment-783445488) and as you can see in services like `plivo` and `whatsapp` we are using [vektra/mockery](https://github.com/vektra/mockery) for generating mocks and writing tests based on those.

For an example on this, please take a look [here](https://github.com/nikoksr/notify/tree/main/service/plivo).

**Review by nikoksr — CHANGES_REQUESTED** [review]

File: go.mod [hunk 7]
```diff
@@ -1,4 +1,4 @@
-module github.com/nikoksr/notify
```
[the remark this answers is not in the document]
Comment (doersa) [reply go.mod:1]: I'm sorry for that, i have a project uses the dingTalk, and the github.com/nikoksr/notify was not accept the PR. i will fix it.

File: service/dingding/README.md (hunk 1 above)
Comment (doersa) [reply service/dingding/README.md]: I'll add readme.md for English.

File: service/dingding/dingding_test.go (hunk 6 above)
Comment (doersa) [reply service/dingding/dingding_test.go:30]: I am not use Interface in dingding, the vektra/mockery is not suitable for this.

**Comment by doersa** [comment]
I commit it again.

**Comment by doersa** [comment]
@nikoksr hi, could the pr is merged?

**Comment by nikoksr** [comment]
Hi @doersa, I'm very sorry for the big delay on my side! Of course, I would like to merge your PR. My only issue is the test file `service/dingding/dingding_test.go`. At the time of writing this Notify is still not at the level of testing that I want it to be at. I'm planning on taking some dedictated time soon to introduce alot more consistency to the repo; meaning, the same doc files for every service and the same depth of testing for every service.

Long story short, I'll leave this open for you do decide. Either implement tests for your `DingDing service` like e.g. the [WhatsApp service](https://github.com/nikoksr/notify/tree/main/service/whatsapp) does it using [vektra/mockery](https://github.com/vektra/mockery), or, remove the test file completely and leave this open to me/the community to fix.

I appreciate your contribution either way and sorry again for the inconvenience!

**Comment by doersa** [comment]
Hi @nikoksr, I don't kwon how to implement the tests with [vektra/mockery](https://github.com/vektra/mockery), so i have deleted the test file completely. 

Best Regards!

**Review by nikoksr — APPROVED** [review]
